secure_tunneling/tests/main.cpp

```diff
@@ -0,0 +1,326 @@
+
+    builderSource.WithOnConnectionShutdown([&]() { fprintf(stdout, "Source Connection Shutdown\n"); });
+
+    builderDestination.WithOnStopped([&](SecureTunnel *secureTunnel) {
+        (void)secureTunnel;
+        fprintf(stdout, "Destination entered Stopped State\n");
+        promiseDestinationStopped.set_value();
+    });
+
+    builderSource.WithOnStopped([&](SecureTunnel *secureTunnel) {
+        (void)secureTunnel;
+        fprintf(stdout, "Source has entered Stopped State\n");
+        promiseSourceStopped.set_value();
+    });
+
+    /* Create Secure Tunnel using the options set with the builder */
+    std::shared_ptr<SecureTunnel> secureTunnelDestination = builderDestination.Build();
+
+    if (!secureTunnelDestination)
+    {
+        fprintf(stderr, "Secure Tunnel Destination Creation failed: %s\n", ErrorDebugString(LastError()));
+        exit(-1);
+    }
+
+    std::shared_ptr<SecureTunnel> secureTunnelSource = builderSource.Build();
+    if (!secureTunnelSource)
+    {
+        fprintf(stderr, "Secure Tunnel Source Creation failed: %s\n", ErrorDebugString(LastError()));
+        exit(-1);
+    }
+
+    fprintf(stdout, "Secure Tunnels Created\n");
+
+    /* Set the Secure Tunnel Client to desire a connected state */
+
+    if (secureTunnelDestination->Start())
+    {
+        fprintf(stderr, "Secure Tunnel Destination Connect call failed: %s\n", ErrorDebugString(LastError()));
+        exit(-1);
+    }
+
+    promiseDestinationConnected.get_future().wait();
+
+    if (secureTunnelSource->Start())
+    {
+        fprintf(stderr, "Secure Tunnel Source Connect call failed: %s\n", ErrorDebugString(LastError()));
+        exit(-1);
+    }
+
+    promiseSourceConnected.get_future().wait();
+
+    promiseDestinationStreamStarted.get_future().wait();
+
+    secureTunnelSource->SendConnectionStart(m_serviceId.value(), connectionId2);
+
+    promiseDestinationConnectionStarted.get_future().wait();
+
+    std::shared_ptr<Message> message1 =
+        std::make_shared<Message>(ByteCursorFromCString(aws_string_c_str(SECTUN_PAYLOAD_MESSAGE)));
+    message1->withServiceId(m_serviceId.value());
+    message1->withConnectionId(connectionId2);
+    secureTunnelSource->SendMessage(message1);
+    fprintf(stdout, "Source Client Sent Message\n");
+
+    promiseDestinationReceivedMessage.get_future().wait();
+
+    std::shared_ptr<Message> message2 =
+        std::make_shared<Message>(ByteCursorFromCString(aws_string_c_str(SECTUN_PAYLOAD_MESSAGE)));
+    message2->withServiceId(m_serviceId.value());
+    message2->withConnectionId(connectionId);
+    secureTunnelDestination->SendMessage(message2);
+    fprintf(stdout, "Destination Client Sent Message\n");
+
+    promiseSourceReceivedMessage.get_future().wait();
+
+    if (secureTunnelDestination->Stop() == AWS_OP_ERR)
+    {
+        fprintf(stderr, "Secure Tunnel Destination Stop call failed: %s\n", ErrorDebugString(LastError()));
+        exit(-1);
```

Comment:
You probably still want to cleanup the buffer on exit, as the buffer will set as long the connection success.
```
    /* Clean Up */
    aws_byte_buf_clean_up(&m_serviceIdStorage);
```